firedrake/pointcloud.py

```diff
@@ -0,0 +1,283 @@
+from firedrake.mesh import spatialindex
+from firedrake.dmplex import build_two_sided
+from firedrake.utils import cached_property
+from firedrake.petsc import PETSc
+from firedrake import logging
+from pyop2.mpi import MPI
+from pyop2.datatypes import IntType, ScalarType
+import numpy as np
+
+
+__all__ = ["PointCloud"]
+
+
+def syncPrint(*args, **kwargs):
+    """Perform a PETSc syncPrint operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncPrint(*args, **kwargs)
+
+
+def syncFlush(*args, **kwargs):
+    """Perform a PETSc syncFlush operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncFlush(*args, **kwargs)
+
+
+class PointCloud(object):
+    """Store points for repeated location in a mesh. Facilitates lookup and evaluation
+    at these point locations.
+    """
+    def __init__(self, mesh, points, *args, **kwargs):
+        """Initialise the PointCloud.
+
+        :arg mesh: A mesh object.
+        :arg points: An array of points.
```

Comment:
What's a point? I think you mean an N x mesh.geometric_dimension() array of point locations.